databricks/koalas/frame.py

```diff
@@ -2950,6 +2950,87 @@ class  locomotion
 
+    def between_time(
+        self, start_time, end_time, include_start=True, include_end=True, axis: Union[int, str] = 0,
+    ) -> "DataFrame":
+        """
+        Select values between particular times of the day (e.g., 9:00-9:30 AM).
+        Parameters
+        ----------
+        start_time : datetime.time or str
+            Initial time as a time filter limit.
+        end_time : datetime.time or str
+            End time as a time filter limit.
+        include_start : bool, default True
+            Whether the start time needs to be included in the result.
+        include_end : bool, default True
+            Whether the end time needs to be included in the result.
+        axis : int, default 0 or 'index'
+            Can only be set to 0 at the moment.
+        Returns
+        -------
+        DataFrame
+            Data from the original dataframe filtered to the specified dates range.
+        Raises
+        ------
+        TypeError
+            If the datatype of index is not datetime
+        Examples
+        --------
+        >>> idx = pd.date_range('2018-04-09', periods=4, freq='1D20min')
+        >>> ts = ks.DataFrame({'A': [1, 2, 3, 4]}, index=idx)
+        >>> ts
+                             A
+        2018-04-09 00:00:00  1
+        2018-04-10 00:20:00  2
+        2018-04-11 00:40:00  3
+        2018-04-12 01:00:00  4
+
+        >>> ts.between_time('0:15', '0:45')
+                             A
+        2018-04-10 00:20:00  2
+        2018-04-11 00:40:00  3
+
+        You get the times that are *not* between two times by setting
+        ``start_time`` later than ``end_time``:
+
+        >>> ts.between_time('0:45', '0:15')
+                             A
+        2018-04-09 00:00:00  1
+        2018-04-12 01:00:00  4
+        """
+        axis = validate_axis(axis)
+        if axis != 0:
+            raise NotImplementedError("between_time currently only works for axis=0")
+
+        if not self.index.is_all_dates:
+            raise TypeError("Datatype of Index must be datetime")
+
+        index_locations = self._get_index_locations(
+            start_time, end_time, include_start, include_end
+        )
+        return self.copy().take(index_locations)
+
+    def _get_index_locations(self, start_time, end_time, include_start=True, include_end=True):
+        """
+        Return index locations of values between particular times of day
+        (e.g., 9:00-9:30AM).
+        Parameters
+        ----------
+        start_time, end_time : datetime.time, str
+            Time passed either as object (datetime.time) or as string in
+            appropriate format ("%H:%M", "%H%M", "%I:%M%p", "%I%M%p",
+            "%H:%M:%S", "%H%M%S", "%I:%M:%S%p","%I%M%S%p").
+        include_start : bool, default True
+        include_end : bool, default True
+        Returns
+        -------
+        values_between_time : array of integers
+        """
+        return self.index.to_pandas().indexer_between_time(
```

Comment:
@shril can we avoid calling `to_pandas()`? It will bring all data from other nodes to the single client node which can easily OOM.

Reply:
@HyukjinKwon, my idea here was to convert just the index column to pandas since there is API support for Datetime index hence the `indexer_between_time()` function.

If you want me to, I can also start with Datetime index support in koalas, and then work on this PR.

Reply:
Thanks for the contribution , @shril . :D

Yeah, I'd say we need Datetime index support in Koalas, not use pandas'.

Reply:
@itholic , starting with a new PR for `DatetimeIndex`. 
@HyukjinKwon can we leave this PR open till I implement the other PR?
Thanks. :)

Reply:
I am fairly new to this so I am sorry if I sound too naive, but doesn't `is_all_dates` function from `koalas` provide support for `Datetimeindex`? @itholic

Reply:
@shril,
Sure, we can just leave this PR as it is and let's revisit after finishing `DatetimeIndex` ! :D

@vmdhhh, 
Thanks for your interest to Koalas !!
`DatetimeIndex` in Koalas now actually not `DatetimeIndex`, but just `Index`.
It's only shown as `DatetimeIndex` only when calling repr as below.
```python
>>> idx = ks.Index([datetime(2019, 1, 1, 0, 0, 0), datetime(2019, 2, 3, 0, 0, 0)])
>>> idx  # It's repr is `DatetimeIndex` since we internally convert this to pandas and use pandas' repr.
DatetimeIndex(['2019-01-01', '2019-02-03'], dtype='datetime64[ns]', freq=None)
>>> type(idx)  # Actually it's instance of `Index`, not `DatetimeIndex`
<class 'databricks.koalas.indexes.Index'>
>>> type(idx.to_pandas())
<class 'pandas.core.indexes.datetimes.DatetimeIndex'>
```

`is_all_dates` just check if all data included in the Index is Spark `TimestampType` type.

https://github.com/databricks/koalas/blob/b81afcc7a461eaad6489e51200862ecdadd04b3c/databricks/koalas/indexes.py#L1973

Reply:
Thank you @itholic for clearing this.